File: nodejs/cronmonitor.js
#!/usr/bin/env node
/**
 * CronMonitor Client for Node.js
 * 
 * A reusable client class for CronMonitor integration.
 * No external dependencies - uses built-in https module.
 * 
 * Usage:
 *   const { CronMonitor } = require('./cronmonitor');
 *   
 *   const monitor = new CronMonitor('YOUR_TOKEN');
 *   
 *   // Option 1: Manual ping
 *   try {
 *       await doWork();
 *       await monitor.ping();
 *   } catch (error) {
 *       process.exit(1);
 *   }
 *   
 *   // Option 2: Wrap your function
 *   await monitor.wrap(async () => {
 *       await doWork();
 *   });
 */

const https = require('https');

class CronMonitor {
    /**
     * Create a CronMonitor client
     * 
     * @param {string} token - Your CronMonitor ping token
     * @param {Object} options - Configuration options
     * @param {number} options.timeout - Request timeout in milliseconds (default: 10000)
     * @param {number} options.retries - Number of retry attempts (default: 3)
     * @param {boolean} options.verbose - Whether to log status messages (default: false)
     */
    constructor(token, options = {}) {
        this.token = token;
        this.timeout = options.timeout || 10000;
        this.retries = options.retries || 3;
        this.verbose = options.verbose || false;
        this.baseUrl = 'cronmonitor.app';
        this.pingPath = `/ping/${token}`;
    }
    
    /**
     * Log message if verbose mode is enabled
     * @private
     */
    _log(message) {
        if (this.verbose) {
            const timestamp = new Date().toISOString();
            console.log(`[${timestamp}] [CronMonitor] ${message}`);
        }
    }
    
    /**
     * Sleep for specified milliseconds
     * @private
     */
    _sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
    
    /**
     * Send a single ping request
     * @private
     */
    _sendRequest() {
        return new Promise((resolve) => {
            const options = {
                hostname: this.baseUrl,
                port: 443,
                path: this.pingPath,
                method: 'GET',
                timeout: this.timeout,
                headers: {
                    'User-Agent': 'CronMonitor-Node/1.0'
                }
            };
            
            const req = https.request(options, (res) => {
                resolve({
                    success: res.statusCode === 200,
                    statusCode: res.statusCode
                });
            });
            
            req.on('error', (err) => {
                resolve({
                    success: false,
                    error: err.message
                });
            });
            
            req.on('timeout', () => {
                req.destroy();
                resolve({
                    success: false,
                    error: 'Request timeout'
                });
            });
            
            req.end();
        });
    }
    
    /**
     * Send success ping to CronMonitor
     * 
     * @returns {Promise<boolean>} True if ping was successful
     */
    async ping() {
        for (let attempt = 1; attempt <= this.retries; attempt++) {
            this._log(`Ping attempt ${attempt}/${this.retries}`);
            
            const result = await this._sendRequest();
            
            if (result.success) {
                this._log('Ping sent successfully');
                return true;
            }
            
            this._log(`Attempt ${attempt} failed: ${result.error || `HTTP ${result.statusCode}`}`);
            
            // Wait before retry (exponential backoff)
            if (attempt < this.retries) {
                const delay = 500 * attempt;
                this._log(`Waiting ${delay}ms before retry...`);
                await this._sleep(delay);
            }
        }
        
        this._log('All ping attempts failed');
        return false;
    }
    
    /**
     * Wrap an async function and ping on success
     * 
     * If the function throws an error, no ping is sent.
     * 
     * @param {Function} fn - Async function to wrap
     * @returns {Promise<any>} - Result of the wrapped function
     */
    async wrap(fn) {
        this._log('Starting wrapped function');
        
        const result = await fn();
        
        this._log('Function completed, sending ping...');
        await this.ping();
        
        return result;
    }
}

// =============================================================================
// CONVENIENCE FUNCTIONS
// =============================================================================

/**
 * Quick ping function for simple use cases
 * 
 * @param {string} token - Your CronMonitor token
 * @param {number} timeout - Request timeout in milliseconds
 * @returns {Promise<boolean>}
 */
async function ping(token, timeout = 10000) {
    const monitor = new CronMonitor(token, { timeout });
    return monitor.ping();
}

/**
 * Wrap a function with CronMonitor ping
 * 
 * @param {string} token - Your CronMonitor token
 * @param {Function} fn - Async function to wrap
 * @returns {Promise<any>}
 */
async function withCronMonitor(token, fn) {
    const monitor = new CronMonitor(token);
    return monitor.wrap(fn);
}

// =============================================================================
// EXPORTS
// =============================================================================

module.exports = {
    CronMonitor,
    ping,
    withCronMonitor
};

// =============================================================================
// EXAMPLE USAGE (runs when executed directly)
// =============================================================================

if (require.main === module) {
    async function exampleJob() {
        console.log('Doing some work...');
        await new Promise(r => setTimeout(r, 1000));
        console.log('Work completed!');
        return { success: true };
    }
    
    async function main() {
        // Create client with verbose logging
        const monitor = new CronMonitor('YOUR_TOKEN', {
            timeout: 10000,
            retries: 3,
            verbose: true
        });
        
        console.log('\n=== Example 1: Manual ping ===');
        try {
            await exampleJob();
            await monitor.ping();
        } catch (error) {
            console.error('Job failed:', error.message);
        }
        
        console.log('\n=== Example 2: Wrap function ===');
        try {
            const result = await monitor.wrap(exampleJob);
            console.log('Result:', result);
        } catch (error) {
            console.error('Job failed:', error.message);
        }
        
        console.log('\n=== Examples completed ===');
    }
    
    main().catch(console.error);
}

File: python/cronmonitor.py
#!/usr/bin/env python3
"""
CronMonitor Python Client

A reusable client class for CronMonitor integration.
No external dependencies - uses standard library only.

Usage:
    from cronmonitor import CronMonitor
    
    monitor = CronMonitor("YOUR_TOKEN")
    
    # Option 1: Manual ping
    try:
        do_work()
        monitor.ping()
    except Exception:
        exit(1)
    
    # Option 2: Decorator
    @monitor.wrap
    def my_job():
        do_work()
    
    # Option 3: Context manager
    with monitor:
        do_work()
"""

import urllib.request
import urllib.error
import functools
import time
from typing import Callable, Any, Optional
from datetime import datetime


class CronMonitor:
    """
    CronMonitor client for Python.
    
    Supports three integration patterns:
    1. Manual: call monitor.ping() after successful job
    2. Decorator: @monitor.wrap decorates your function
    3. Context manager: use with 'with' statement
    """
    
    BASE_URL = "https://cronmonitor.app"
    
    def __init__(
        self,
        token: str,
        timeout: int = 10,
        retries: int = 3,
        verbose: bool = False
    ):
        """
        Initialize CronMonitor client.
        
        Args:
            token: Your CronMonitor ping token
            timeout: Request timeout in seconds
            retries: Number of retry attempts
            verbose: Whether to print status messages
        """
        self.token = token
        self.timeout = timeout
        self.retries = retries
        self.verbose = verbose
        self._ping_url = f"{self.BASE_URL}/ping/{token}"
    
    def _log(self, message: str) -> None:
        """Print message if verbose mode is enabled"""
        if self.verbose:
            timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            print(f"[{timestamp}] [CronMonitor] {message}")
    
    def ping(self) -> bool:
        """
        Send success ping to CronMonitor.
        
        Returns:
            True if ping was successful, False otherwise
        """
        for attempt in range(1, self.retries + 1):
            try:
                self._log(f"Ping attempt {attempt}/{self.retries}")
                
                request = urllib.request.Request(
                    self._ping_url,
                    headers={"User-Agent": "CronMonitor-Python/1.0"}
                )
                
                response = urllib.request.urlopen(request, timeout=self.timeout)
                
                if response.status == 200:
                    self._log("Ping sent successfully")
                    return True
                    
            except urllib.error.URLError as e:
                self._log(f"Attempt {attempt} failed: {e}")
            except Exception as e:
                self._log(f"Attempt {attempt} failed: {e}")
            
            # Wait before retry (exponential backoff)
            if attempt < self.retries:
                sleep_time = 0.5 * attempt
                self._log(f"Waiting {sleep_time}s before retry...")
                time.sleep(sleep_time)
        
        self._log("All ping attempts failed")
        return False
    
    def wrap(self, func: Callable) -> Callable:
        """
        Decorator that pings CronMonitor after successful execution.
        
        Usage:
            @monitor.wrap
            def my_job():
                do_something()
        """
        @functools.wraps(func)
        def wrapper(*args, **kwargs) -> Any:
            self._log(f"Starting wrapped function: {func.__name__}")
            
            # Run the actual function
            result = func(*args, **kwargs)
            
            # Ping on success
            self._log(f"Function {func.__name__} completed, sending ping...")
            self.ping()
            
            return result
        
        return wrapper
    
    def __enter__(self) -> "CronMonitor":
        """Context manager entry"""
        self._log("Entering CronMonitor context")
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb) -> bool:
        """
        Context manager exit.
        Pings CronMonitor only if no exception occurred.
        """
        if exc_type is None:
            self._log("Context completed successfully, sending ping...")
            self.ping()
        else:
            self._log(f"Context exited with exception: {exc_type.__name__}")
        
        # Don't suppress exceptions
        return False


# =============================================================================
# CONVENIENCE FUNCTIONS
# =============================================================================

def ping(token: str, timeout: int = 10) -> bool:
    """
    Quick ping function for simple use cases.
    
    Usage:
        from cronmonitor import ping
        
        try:
            do_work()
            ping("YOUR_TOKEN")
        except Exception:
            exit(1)
    """
    return CronMonitor(token, timeout=timeout).ping()


def cronmonitor(token: str, timeout: int = 10) -> Callable:
    """
    Decorator factory for quick decoration.
    
    Usage:
        from cronmonitor import cronmonitor
        
        @cronmonitor("YOUR_TOKEN")
        def my_job():
            do_something()
    """
    monitor = CronMonitor(token, timeout=timeout)
    return monitor.wrap


# =============================================================================
# EXAMPLE USAGE
# =============================================================================

if __name__ == "__main__":
    import sys
    
    # Create monitor instance
    monitor = CronMonitor(
        token="YOUR_TOKEN",
        timeout=10,
        retries=3,
        verbose=True
    )
    
    # Example 1: Manual ping
    print("\n=== Example 1: Manual ping ===")
    try:
        print("Doing some work...")
        time.sleep(0.5)
        print("Work completed!")
        monitor.ping()
    except Exception as e:
        print(f"Failed: {e}")
    
    # Example 2: Decorator
    print("\n=== Example 2: Decorator ===")
    
    @monitor.wrap
    def my_decorated_job():
        print("Running decorated job...")
        time.sleep(0.5)
        print("Decorated job done!")
        return "result"
    
    result = my_decorated_job()
    print(f"Got result: {result}")
    
    # Example 3: Context manager
    print("\n=== Example 3: Context manager ===")
    
    with monitor:
        print("Running in context...")
        time.sleep(0.5)
        print("Context work done!")
    
    print("\n=== All examples completed ===")
